File: examples/csharp/avalonia/NotesShowcase/Models/NoteModel.cs
namespace NotesShowcase.Models;

/// <summary>
/// Domain model for a single note.
///
/// See spec/proposals/2026-05-29-notes-showcase-scenario.md §3.2.
/// Pure data — no behavior, no VMx dependencies.
///
/// VMx v3 (ADR-0048 §2.1): <c>FormVM&lt;NoteModel&gt;.IsDirty</c> derives from the
/// model's own <c>object.Equals</c>, and the default snapshotter is a deep
/// System.Text.Json round-trip. A record's synthesized equality compares the
/// <see cref="Tags"/> collection by <em>reference</em>, so the deep-copied
/// snapshot's fresh list would read as perpetually dirty. The value-equality
/// override below (the documented v3 hook) compares <see cref="Tags"/>
/// element-wise so a structurally-identical snapshot is correctly clean.
/// </summary>
public sealed record NoteModel(
    string Id,
    string NotebookId,
    string Title,
    IReadOnlyList<string> Tags,
    string Body,
    bool Starred,
    DateTimeOffset CreatedAt,
    DateTimeOffset UpdatedAt)
{
    /// <summary>
    /// Value equality with element-wise <see cref="Tags"/> comparison
    /// (replaces the record-synthesized reference comparison for the list).
    /// </summary>
    public bool Equals(NoteModel? other)
    {
        if (other is null) return false;
        if (ReferenceEquals(this, other)) return true;
        return Id == other.Id
            && NotebookId == other.NotebookId
            && Title == other.Title
            && Body == other.Body
            && Starred == other.Starred
            && CreatedAt == other.CreatedAt
            && UpdatedAt == other.UpdatedAt
            && Tags.SequenceEqual(other.Tags);
    }

    /// <inheritdoc/>
    public override int GetHashCode()
    {
        var hash = new HashCode();
        hash.Add(Id);
        hash.Add(NotebookId);
        hash.Add(Title);
        hash.Add(Body);
        hash.Add(Starred);
        hash.Add(CreatedAt);
        hash.Add(UpdatedAt);
        foreach (var tag in Tags)
            hash.Add(tag);
        return hash.ToHashCode();
    }
}

File: examples/swift/notes-showcase/Sources/NotesShowcaseCore/Models/NoteModel.swift
import Foundation

/// Domain model for a single note.
///
/// See spec/proposals/2026-05-29-notes-showcase-scenario.md §3.2.
/// Pure data — no behavior, no VMx dependencies.
///
/// Custom `Equatable` compares `tags` element-wise so that
/// `FormVM<NoteModel>.isDirty` correctly reads clean after a deep snapshot
/// round-trip. (In the C# flavor a record's synthesised equality compares the
/// list by reference; Swift's `[String] ==` is already element-wise, but the
/// explicit conformance makes the intent explicit and mirrors the C# hook.)
public struct NoteModel: Sendable, Codable {
    public let id: String
    public let notebookId: String
    public let title: String
    public let tags: [String]
    public let body: String
    public let starred: Bool
    public let createdAt: Date
    public let updatedAt: Date

    public init(
        id: String,
        notebookId: String,
        title: String,
        tags: [String],
        body: String,
        starred: Bool,
        createdAt: Date,
        updatedAt: Date
    ) {
        self.id = id
        self.notebookId = notebookId
        self.title = title
        self.tags = tags
        self.body = body
        self.starred = starred
        self.createdAt = createdAt
        self.updatedAt = updatedAt
    }

    /// Returns a copy with the specified fields overridden.
    /// Use this to make edits — mirrors the C# `record with { … }` syntax.
    public func with(
        notebookId: String? = nil,
        title: String? = nil,
        tags: [String]? = nil,
        body: String? = nil,
        starred: Bool? = nil,
        updatedAt: Date? = nil
    ) -> NoteModel {
        NoteModel(
            id: self.id,
            notebookId: notebookId ?? self.notebookId,
            title: title ?? self.title,
            tags: tags ?? self.tags,
            body: body ?? self.body,
            starred: starred ?? self.starred,
            createdAt: self.createdAt,
            updatedAt: updatedAt ?? self.updatedAt
        )
    }
}

extension NoteModel: Equatable {
    /// Value equality with element-wise `tags` comparison.
    /// Replaces the default synthesised equality to mirror the C# custom Equals.
    public static func == (lhs: NoteModel, rhs: NoteModel) -> Bool {
        lhs.id == rhs.id
            && lhs.notebookId == rhs.notebookId
            && lhs.title == rhs.title
            && lhs.body == rhs.body
            && lhs.starred == rhs.starred
            && lhs.createdAt == rhs.createdAt
            && lhs.updatedAt == rhs.updatedAt
            && lhs.tags == rhs.tags
    }
}
